library/ssl_tls.c

```diff
@@ -3526,81 +3540,16 @@ static int ssl_parse_record_header( mbedtls_ssl_context *ssl )
         return( MBEDTLS_ERR_SSL_INVALID_RECORD );
     }
 
-    /* Check length against bounds of the current transform and version */
-    if( ssl->transform_in == NULL )
-    {
-        if( ssl->in_msglen < 1 ||
-            ssl->in_msglen > MBEDTLS_SSL_MAX_CONTENT_LEN )
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad message length" ) );
-            return( MBEDTLS_ERR_SSL_INVALID_RECORD );
-        }
-    }
-    else
-    {
-        if( ssl->in_msglen < ssl->transform_in->minlen )
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad message length" ) );
-            return( MBEDTLS_ERR_SSL_INVALID_RECORD );
-        }
-
-#if defined(MBEDTLS_SSL_PROTO_SSL3)
-        if( ssl->minor_ver == MBEDTLS_SSL_MINOR_VERSION_0 &&
-            ssl->in_msglen > ssl->transform_in->minlen + MBEDTLS_SSL_MAX_CONTENT_LEN )
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad message length" ) );
-            return( MBEDTLS_ERR_SSL_INVALID_RECORD );
-        }
-#endif
-#if defined(MBEDTLS_SSL_PROTO_TLS1) || defined(MBEDTLS_SSL_PROTO_TLS1_1) || \
-    defined(MBEDTLS_SSL_PROTO_TLS1_2)
-        /*
-         * TLS encrypted messages can have up to 256 bytes of padding
-         */
-        if( ssl->minor_ver >= MBEDTLS_SSL_MINOR_VERSION_1 &&
-            ssl->in_msglen > ssl->transform_in->minlen +
-                             MBEDTLS_SSL_MAX_CONTENT_LEN + 256 )
-        {
-            MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad message length" ) );
-            return( MBEDTLS_ERR_SSL_INVALID_RECORD );
-        }
-#endif
-    }
-
     /*
-     * DTLS-related tests done last, because most of them may result in
-     * silently dropping the record (but not the whole datagram), and we only
-     * want to consider that after ensuring that the "basic" fields (type,
-     * version, length) are sane.
+     * DTLS-related tests.
+     * Must be done before checking the bounds for the length
+     * field because the latter depends on the current epoch.
```

Comment:
I don't see where that dependency is when reading the code. Could you clarify which length depends on the epoch and how?

Reply:
@gilles-peskine-arm See IOTSSL-1417. The minimal record length depends on the negotiated cipher suite and hence changes with the epoch; during the first handshake, it's `1`, afterwards it's larger because of the presence of a MAC e.g. The reason I swapped the code-paths is that records with an old epoch are silently discarded without skipping the datagram, while records that are too short lead to dismissal of the entire datagram. This leads to a problem in the following laboratory test-case: The `ChangeCipherSpec` message from the first handshake got duplicated and merged with the subsequent `Finished` message. After the first `ChangeCipherSpec`, the minimal record length increases to `> 1`, which renders the duplicated `ChangeCipherSpec` too short and therefore leads to dismissal of the entire datagram, including the `Finished` message. In contrast, swapping the length and epoch checks leads to the duplicated `ChangeCipherSpec` being silently dropped, and the `Finished` message is parsed as expected.

It's a long time ago that I worked on this PR. Perhaps we should consult the DTLS RFC to see if it mandates a specific order of checks, and whether such strange laboratory scenarios as the one just described are potentially excluded.

I came across that issue while playing with the UDP proxy and UDP packer - duplication and packing in conjunction lead to the above problem quite quickly.

Reply:
Thanks for the explanation, please put it in a comment.

Reply:
@mpg I didn't find anything in the standard saying that some failures should lead to dropping an entire datagram, while others should only lead to discarding the record. The only relevant section seems to be [this one](https://tools.ietf.org/html/rfc6347#section-4.1.2.7). Do you know more about that?

Reply:
@gilles-peskine-arm Done.